src/libawkward/array/IndexedArray.cpp

```diff
@@ -38,6 +38,11 @@ namespace awkward {
     return ISOPTION;
   }
 
+  template <typename T, bool ISOPTION>
+  bool IndexedArrayOf<T, ISOPTION>::isindexed() const {
+    return true;
+  }
+
```

Comment:
> Okay, so that's what `isindexed` means. I guess there are five cases to `dynamically_cast`, so it's not purely a performance consideration. Still, if we're going to introduce booleans like this, I'd like to do it more systematically (e.g. "index-like," "list-like," "record-like," etc.).
> 
Yes, it would be a Python-like approach to a Python-like `isinstance()`

> Also, after determining it's an IndexedArray, don't you have to call a non-virtual function immediately afterward? To do that, don't you have to do a `dynamic_cast` right away anyway?
> 
No, since I have to explicitly avoid doubling-up the OptionType.

> Maybe a more "systematic solution" would be to make all IndexedArray specializations inherit from a pure-virtual IndexedArray class below `Content`? I avoided deep inheritance trees because that mistake has bitten me in the past, but this could be a well-motivated three-level structure (IndexedOptionArray64 < IndexedArrayBase < Content). It could replace things like
> 
> ```c++
> if (IndexedArray32* raw = dynamic_cast<IndexedArray32*>(content_.get())) {
>     raw->do_same_thing();
> }
> else if (IndexedArrayU32* raw = dynamic_cast<IndexedArrayU32*>(content_.get())) {
>     raw->do_same_thing();
> }
> else if (IndexedArray64* raw = dynamic_cast<IndexedArray64*>(content_.get())) {
>     raw->do_same_thing();
> }
> else if (IndexedOptionArray32* raw = dynamic_cast<IndexedOptionArray32*>(content_.get())) {
>     raw->do_same_thing();
> }
> else if (IndexedOptionArray64* raw = dynamic_cast<IndexedOptionArray64*>(content_.get())) {
>     raw->do_same_thing();
> }
> else {
>     deal_with_non_indexed_array();
> }
> ```
> 
> with
> 
> ```c++
> if (IndexedArrayBase* raw = dynamic_cast<IndexedArrayBase*>(content_.get())) {
>     raw->do_same_thing();
> }
> else {
>     deal_with_non_indexed_array();
> }
> ```
> 
> which would be a valuable simplification.

Yes, I like a more "systematic solution".

Reply:
To avoid making IndexedArrays of IndexedArrays, we should write the simplest code and then call `outputIndexedArray.simplify()`, which collapses IndexedArrays of IndexedArrays. That might also avoid branching—we might be able to do the same thing whether the content is an IndexedArray or not.

The two simplification routines, `IndexedArray::simplify` and `UnionArray::simplify`, are complex with lots of `dynamic_casts` because in general, you need to do that. (`UnionArray::simplify`'s casting is in some of the `merge` implementations.) The reason I wrote them is to concentrate all that special-casing so that we don't have to do it again.

Even if creating an extra IndexedArray or UnionArray and then calling `simplify` to get rid of it is a little slower (less than a factor of 2 or so), it's preferable for keeping the code manageable. We'll be adding a lot of operations, potentially forever, so we have to make sure that adding new ones doesn't scale up the complexity of the project. (I.e. they should become routine.)

Reply:
yes, this would definitely help to remove the branching

Reply:
removed as not needed